src/crimson/admin/pg_commands.cc

```diff
@@ -0,0 +1,159 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#include "crimson/admin/pg_commands.h"
+
+#include <memory>
+#include <string>
+#include <string_view>
+
+#include <fmt/format.h>
+#include <seastar/core/future.hh>
+
+#include "crimson/admin/admin_socket.h"
+#include "crimson/osd/osd.h"
+#include "crimson/osd/pg.h"
+
+
+using crimson::osd::OSD;
+using crimson::osd::PG;
+using namespace crimson::common;
+
+
+namespace crimson::admin::pg {
+
+class PGCommand : public AdminSocketHook {
+public:
+  // TODO: const correctness of osd
+  PGCommand(crimson::osd::OSD& osd,
+            std::string_view prefix,
+            std::string_view desc,
+            std::string_view help)
+      : AdminSocketHook{prefix, desc, help}, osd {osd}
+  {}
+  seastar::future<tell_result_t> call(const cmdmap_t& cmdmap,
+                                      std::string_view format,
+                                      ceph::bufferlist&& input) const final
+  {
+    // we have "ceph tell <pgid> <cmd>". and it is the ceph cli's responsibility
+    // to add "pgid" to the cmd dict. as rados_pg_command() does not set it for
+    // us. moreover, and "pgid" is not listed in the command description, as user
+    // command format does not follow the convention of "<prefix> [<args>,...]"
+    // so we have to verify it on the server side.
+    std::string pgid_str;
+    pg_t pgid;
+    if (!cmd_getval(cmdmap, "pgid", pgid_str)) {
+      return seastar::make_ready_future<tell_result_t>(
+        tell_result_t{-EINVAL, "no pgid specified"});
+    } else if (!pgid.parse(pgid_str.c_str())) {
+      return seastar::make_ready_future<tell_result_t>(
+        tell_result_t{-EINVAL, fmt::format("couldn't parse pgid '{}'", pgid_str)});
+    }
+    // am i the primary for this pg?
```

Comment:
I would suggest that if we encounter this check ("am I the primary") a second time, this block
of checks should be factored out. Not urgent.

Reply:
could you be more specific? what change do you suggest? `PGCommand` tries to extract the shared code among pg tell commands. what kind of "second time" we are expecting?

Reply:
Sure. I was ref the code that starts with a pgid, and:
- fetches the osd map,
- verifies that the named pg exists (according to the map), and
- that we are part of the PG's set, and
- we are the primary,

returning either a ref<pg> or an error.

I suspect that this is a common test, and we may want to have it as separate method (not necessarily returning a tell_t, as that indeed would limit applicability).

Reply:
Thanks Ronen, I understand you concern now. But I am not sure how to address it with a balanced way between “don’t repeat yourself” and too much indirection layers. As this is two lines of code..

Could you sent a follow-up PR to help me get this right? Much appreciated.

Reply:
WIll do - the minute I encounter a second usage of that...
More seriously - will indeed do that once I return to Crimson code.